File: src/2_addTwoNumbers/Solution.cpp
#include <string>
#include <iostream>

using namespace std;
/**
 * Definition for singly-linked list.
*/
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int q = 0;
        ListNode *p = NULL;
        ListNode *rst = NULL;

        while (l1 != NULL && l2 != NULL) {
            int x = l1->val + l2->val + q;
            if (x >= 10) { x -= 10; q = 1; }
            else q = 0;

            ListNode *next = new ListNode(x);
            if (p) { p->next = next; }
            else { rst = next; }
            p = next;
            l1 = l1->next; l2 = l2->next;
        }

        while (l1 != NULL) {
            int x = q + l1->val;
            if (x >= 10) { x -= 10; q = 1; }
            else q = 0;

            ListNode *next = new ListNode(x);
            if (p) { p->next = next; }
            else { rst = next; }
            p = next;
            l1 = l1->next;
        }
        while (l2 != NULL) {
            int x = q + l2->val;
            if (x >= 10) { x -= 10; q = 1; }
            else q = 0;

            ListNode *next = new ListNode(x);
            if (p) { p->next = next; }
            else { rst = next; }
            p = next;
            l2 = l2->next;
        }
        if (q == 1)
            p->next = new ListNode(q);

        return rst;
    }
};

int main() {
    int x; 

    ListNode *l1, *l2, *p;
    p = l1 = l2 = NULL;

    while (cin >> x) {
        if (x == -1) break;
        ListNode *node = new ListNode(x);
        if (!l1) { l1 = node; }
        else { p->next = node; }
        p = node;
    }
    p = NULL;
    while (cin >> x) {
        ListNode *node = new ListNode(x);
        if (!l2) { l2 = node; }
        else { p->next = node; }
        p = node;
    }

    Solution s;
    p = s.addTwoNumbers(l1, l2);

    while (p->next) {
        cout << p->val << "->";
        p = p->next;
    }
    cout << p->val << endl;
}

File: src/2_addTwoNumbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        else if (l2 == null) return l1;

        int sum  = l1.val + l2.val;
        int q = 0;
        if (sum >= 10) { sum -= 10; q = 1; }
        ListNode lsum = new ListNode(sum);
        ListNode l;
        l1 = l1.next; l2 = l2.next;
        for (l = lsum; l1 != null && l2 != null; l1 = l1.next, l2 = l2.next) {
            int s = l1.val + l2.val + q;
            if (s >= 10) { s -= 10; q = 1; }
            else q = 0;
            l.next = new ListNode(s);
            l = l.next;
        }

        while (l1 != null) {
            int s = l1.val + q;
            if (s >= 10) { s -= 10; q = 1; }
            else q = 0;
            l.next = new ListNode(s);
            l1 = l1.next;
            l = l.next;
        }
        while (l2 != null) {
            int s = l2.val + q;
            if (s >= 10) { s -= 10; q = 1; }
            else q = 0;
            l.next = new ListNode(s);
            l2 = l2.next;
            l = l.next;
        }
        if (q == 1) {
            l.next = new ListNode(1);
        }
        return lsum;
    }
}
